### PR TITLE
Gutenlypso: Copy Post copies the most recent version of the post

### LINKED ISSUE #29932: Gutenlypso: Impossible to duplicate drafts
The Copy Post feature doesn't work on drafts created during the current browser session.

This happens because the copied post is fetched at `mapStateToProps` time, and added to the new post's `initialEdits` prop:

https://github.com/Automattic/wp-calypso/blob/30f5cf5d5de1a2795bbd3d7301e7da37c963de40/client/gutenberg/editor/main.jsx#L207

using the `requestSitePost` data getter:

https://github.com/Automattic/wp-calypso/blob/30f5cf5d5de1a2795bbd3d7301e7da37c963de40/client/gutenberg/editor/main.jsx#L150-L157

Such function keeps the response cached indefinitely with a "siteId-postId" key.
The cache can be invalidated passing a `freshness = 0` argument, which then forces calling the API again:

https://github.com/Automattic/wp-calypso/blob/30f5cf5d5de1a2795bbd3d7301e7da37c963de40/client/state/data-getters/index.js#L208-L234

In our case, this is roughly what happens:

1. Start a new draft.
2. As soon as the post gets an ID (let's say `1`), `requestSitePost` is called for post `1`, returning a roughly empty post object, which then gets cached for post `1`.
3. Save this draft, and navigate back to the drafts list.
4. From the ellipsis menu of draft `1` item, select "Duplicate".
5. A new draft is started, this time with ID `2`.
6. `requestSitePost` is eventually called for post `2`, and it's empty response gets cached for post `2`.
7. Then, in `mapStateToProps` we request post `1` to obtain the data to add to `2` initial edits. This hits the cache set in step 2, which is an empty post object.
8. The empty post object is added to `2` initial edits, which means `1` wasn't duplicated at all.

---

I've tried adding a freshness to the `copyPost` request, but it requires moving it somewhere else, because of course calling an uncached API in `mapStateToProps` inevitably crashes the browser.

---

Props to @mmtr for the excellent catch: https://github.com/Automattic/wp-calypso/pull/29911#pullrequestreview-189770314

### PR DESCRIPTION
**Pull request by Copons**
#### Changes proposed in this Pull Request

* Invalidate the post API cache before requesting the post to duplicate.

Previously, the duplicated post was being requested via a simple `requestSitePost` called from the editor Main component:

https://github.com/Automattic/wp-calypso/blob/30f5cf5d5de1a2795bbd3d7301e7da37c963de40/client/gutenberg/editor/main.jsx#L207

`requestSitePost` caches the result indefinitely, and it's called as the first thing when opening an editor instance, and NOT after the post is edited.

If the post to duplicate were created (or updated) in the same session, that `requestSitePost` would contain a version of the post preceding the latest changes.
In case of a newly created post, for example, the version would be (roughly) empty.

Now, if we're duplicating a post, the first thing we do is calling a `requestSitePost` with 0 freshness, which replaces the cached version with a fresh one.
Then, when we enter the editor, the following `requestSitePost` calls will hit the fresh cache instead of the stale one.

#### Testing instructions

* Open `/block-editor`.
* Add the title and save it as a draft.
* Navigate back to the posts list.
* Open the ellipsis menu of the draft, and click Duplicate.
* ⚠️ The new post should contain the copied post title.
* Navigate back to the posts list and edit another draft post.
* Modify its title and save it as a draft.
* Navigate back to the posts list.
* Open the ellipsis menu of this other draft, and click Duplicate.
* ⚠️ The new post should contain the draft title, updated with the latest changes.

Fixes #29932

### CODE REVIEW

File: client/gutenberg/editor/controller.js [hunk 1]
```diff
@@ -152,6 +156,12 @@ export const post = async ( context, next ) => {
 		const siteId = getSelectedSiteId( state );
 		const userId = getCurrentUserId( state );
 
+		// When copying a post, first invalidate the cache containing the `duplicatePostId` post,
+		// so that following `requestSitePost` will hit a freshly cached version of the post.
+		if ( !! duplicatePostId ) {
+			requestSitePost( siteId, duplicatePostId, postType, 0 );
```
Comment (mmtr) [inline client/gutenberg/editor/controller.js]: I understand this logic thanks to the great comment and because I have the context of the PR, but it seems to me a bit weird to request a post which we're not using at all, just because this will clear the cache for the next request. 

Why don't we request the non-cached version of the post directly here?

https://github.com/Automattic/wp-calypso/blob/30f5cf5d5de1a2795bbd3d7301e7da37c963de40/client/gutenberg/editor/main.jsx#L207

We can add a new `freshness` parameter to `getPost` that can be sent to `requestSitePost`:

```js
const getPost = ( siteId, postId, postType, freshness ) => {
	if ( siteId && postId && postType ) {
		const requestSitePostData = requestSitePost( siteId, postId, postType, freshness );
		return get( requestSitePostData, 'data', null );
	}

	return null;
};
```

So we can tell `getPost` we want the fresh post data:

```js
const postCopy = getPost( siteId, duplicatePostId, postType, 0 );
```

Comment (Copons) [reply client/gutenberg/editor/controller.js]: That was my first thought as well, but unfortunately, that `getPost` method is called inside `mapStateToProps`, which means that it's called _a lot_ of times.

It's fine enough to do so if we hit the cache, as it's pretty much the same of a Redux selector.
But if we force the freshness to 0, then we would actually fire an actual network request every time the state changes, which ultimately results in crashing the browser.

I totally understand when you say that it's a confusing logic, but at the same time, it's very similar to our own Query Component pattern, where we force an API call somewhere (usually in the parent component), and use the data somewhere else (usually in the children).

I also tried to only call the 0-freshness `getPost` once in the Main, but introducing an internal state for it sounds more painful (and would waste more re-renders) than the proposed solution.

Comment (mmtr) [reply client/gutenberg/editor/controller.js]: > But if we force the freshness to 0, then we would actually fire an actual network request every time the state changes, which ultimately results in crashing the browser.

Indeed we would make a request, but only if we're duplicating a post, which I guess is what we want. `getPost` doesn't call to `requestSitePost` if `duplicatePostId` is not truthy.

Or am I missing something? Does the state still change a lot when `duplicatePostId` has a value?

Comment (Copons) [reply client/gutenberg/editor/controller.js]: When we copy a post, `duplicatePostId` is passed as query param, and so it's always truthy from when the page starts loading until when the draft is saved and the URL is automatically updated.
Though, that's unrelated to how many times the `mapStateToProps`, and consequently `requestSitePost`, is called when the editor load.
`mapStateToProps` is called every time the Redux state changes, regardless of the `ownProps` argument values.

I've just tried just to be sure.
On copy post, the Main's `mapStateToProps` is called 13 times when we enter the editor, 5 times when we leave it, and 1 every time we do some header actions (e.g. save, publish, preview, etc.).

In other words, we would end up calling the posts API 13 times in roughly a second, and several other - completely unnecessary - times while we work on the post.

Comment (mmtr) [reply client/gutenberg/editor/controller.js]: Thanks for confirming!
